### DISCUSSION TITLE
Do we need to save id in splay tree?

### DISCUSSION THREAD

**Discussion by euph0nium** [ImageMagick/ImageMagick#4724] [Development] [answered]
```c++
/* wand.c AcquireWandId() */ 
 if (wand_ids == (SplayTreeInfo *) NULL)
    wand_ids=NewSplayTree((int (*)(const void *,const void *)) NULL,
      (void *(*)(void *)) NULL,(void *(*)(void *)) NULL);
  wand_id=id++;
  (void) AddValueToSplayTree(wand_ids,(const void *) wand_id,
    (const void *) wand_id);
```
In `AcquireWandId()` func,  we define static variable `id`. And returned id will be `id+1`.
```c++
/* wand.c RelinquishWandId() */
 if (wand_ids != (SplayTreeInfo *) NULL)
    (void) DeleteNodeFromSplayTree(wand_ids,(const void *) id);
```
In `RelinquishWandId()` func, we delete the id from splay tree.
Since we don't recycle  relinquished id, do we need to save id in splay tree?

**Comment by urban-warrior** [comment]
The Splay tree stores the wand ID as an address, in this case 1, 2, 3, 4, ....  This works because there are no destructors defined in NewSplayTree().  Instead, when deleting a tree node, we use the address to locate the node we want to toss.  Do you need clarity?  Or do you think you found a flaw in tracking Wand ID's?

**Reply by euph0nium** [reply]
Thanks for that. 
I'm learning IM code. Maybe my question is not clear. I want to know the intent of saving ids in a centralized data struct.

**Comment by urban-warrior** [comment, marked as answer]
Each wand you create has a unique id.  We need an address so we can identify it when the wand is destroyed and the wand_id is destroyed as well.  Wand ids are useful for assigning a wand an unique name so they can be tracked by the user or for debugging purposes.

**Reply by euph0nium** [reply]
Thanks. I get it.
